community/rfcs/20200205-weight-clustering.md

```diff
@@ -0,0 +1,185 @@
+    ```
+
+2. **Fine-tuning** &mdash; train the clustered model to fine-tune the initial values of cluster centroids:
+
+   ```python
+   clustered_model.fit(x_training_data, y_training_data, epochs=1)
+   ```
+
+3. **Stripping** &mdash; call `strip_clustering()` to strip the clustering wrappers from the model and re-obtain the original model structure:
+
+   ```python
+   stripped_model = cluster.strip_clustering(clustered_model)
+   ```
+
+The code snippets below illustrate common usage patterns of the clustering API. For the sake of brevity, the clustering examples presented here only deal with the 1st step of the above workflow.
+
+A more comprehensive end-to-end example on how to use the API to cluster and save a model is provided in the examples directory, under *clustering*.
+
+### Example 0: Cluster an Entire Model
+
+When passing an entire model to `cluster_weights()`, the tool will determine which layers are clusterable and will cluster all of them &mdash; other layers will be ignored and left as they are.
+
+```python
+clustering_params = {
+    'number_of_clusters': 8,
+    'cluster_centroids_init': 'linear'
+}
+
+# Clustering a sequential model
+model = keras.Sequential([
+    layers.Dense(10, input_shape=(10,)),
+    layers.Dense(10),
+])
+
+clustered_model = cluster.cluster_weights(model, **clustering_params)
+```
+
+```python
+# Clustering a functional model
+i1 = keras.Input(shape=(10,))
+i2 = keras.Input(shape=(10,))
+x1 = layers.Dense(10)(i1)
+x2 = layers.Dense(10)(i2)
+outputs = layers.Add()([x1, x2])
+
+model = keras.Model(inputs=[i1, i2], outputs=outputs)
+
+clustered_model = cluster.cluster_weights(model, **clustering_params)
+```
+
+### Example 1: Cluster Some of the Layers
+
+Individual layers within a model can be clustered selectively by invoking `cluster_weights()` on a layer object, as follows:
+
+```python
+clustered_model = keras.Sequential()
+clustered_model.add(cluster.cluster_weights(keras.layers.Dense(10), **clustering_params))
+clustered_model.add(keras.layers.Dense(10))
+```
+In the above example model the weights of the 1st `Dense` layer will be clustered, whereas the weights of the 2nd layer will not be clustered.
+
+### Example 2: Export Clustered Model
+
+```python
+clustered_model = cluster_weights(model, **clustering_params)
+keras.models.save_model(clustered_model, keras_file)
+```
+
+### Example 3: Import Clustered Model
+
+A clustered Keras model needs to be loaded within a `cluster_scope` in order to be successfully deserialized:
+
+```python
+with cluster_scope():
+  loaded_model = keras.models.load_model(keras_file)
+```
+
+## Experimental Results
+
+The tables below summarizes the compression rate and top-1 accuracy achieved when clustering the weights of different models using the following parameters:
+
+```python
+{ 'number_of_clusters': 32, 'cluster_centroids_init': 'linear' }
+```
+
+### ConvNet (trained on MNIST)
+
+| Variant        | Size of compressed .tflite file | Top-1 accuracy  |
```

Comment:
Yes. It would also match https://www.tensorflow.org/model_optimization/guide/pruning also.

It'd also be good to mention the number of clusters you used to achieve each of these results (to match with what is done for pruning).

Reply:
The other thing is that the results seem decent but it'd be valuable to have a Mobilenet v2 result that is less aggressively clustered (fewer compression benefits, but higher model accuracy).

The almost 6x compression you got here in weight clustering is close to what we'd get with model pruning with 90% sparsity. People will see your results and compare them to the [Mobilenet v1 results](https://www.tensorflow.org/model_optimization/guide/pruning) for pruning.

An extra Mobilenet v2 data point the minimizes the number of clusters but gets a top-1 accuracy drop by at most 1% would be useful for that. You can even leave additional data points that vary the number of clusters in the comments, even if we don't ultimately show them in the RFC / user documentation.

Reply:
> It'd also be good to mention the number of clusters you used to achieve each of these results

The number of clusters is 32 for all these examples, as described in the introductory paragraph of the _Experimental Results_ section. Is that too vague? Should I add the number of clusters explicitly to each table?

Reply:
> An extra Mobilenet v2 data point that minimizes the number of clusters but still gets a top-1 accuracy drop by at most 1% would be useful for that.
We are working on that and include the data into the overview document.

More generally, what is the intention for the RFC? Is it supposed to be updated as we keep improving both the technique/implementation and the experimental results? If a significant change is introduced, do we create a new RFC, or update the existing one? The name seems to imply the former. Also, since there is the "Results" section in the overview document, it appears to be a natural place to keep track of changes in experimental results.